WordPress/gutenberg
# Rich Text: Formalize RichText children value abstraction

Related: #7476 , #7890 

This pull request seeks to work toward detaching the dependency between RichText value and the `@wordpress/element` React element shape. It formalizes what we already offer in the `children` source as a proper value shape of its own.

**Implementation notes:**

Some observations made in the implementation herein:

- "Element" is an implementation detail. The only requirements of our RichText implementation is being able to source from and write to HTML. Representing the value as an object shape (vs. a plain markup string) has been the subject of much past discussion (#5380, #2750) where the conclusion is that we want some of the benefits which could be offered through element (e.g. safety, content analysis) but it does not strictly need to be element itself. Our current need for manipulation is concatenation (merging), which can be offered through an abstraction native to the "children" type.
   - Our need for `dom-react` hinges on this fact that we expect to need to convert a DOM node back to an element for serialization. But our custom serializer (#5897) is more tolerant to non-"element" props (e.g. `class`, `style`). It is fine if this becomes non-true in the future, so long as we have the abstractions in place for `children` and `node` source types to work from and to the HTML.
- This **does not intend to propose any specific object shape for the RichText value**. Initial explorations here did, and they introduced needless complexity, when the most critical piece is providing the consistent abstraction. I'll leave value shape to efforts like in #7890 (issue at #771).
- The implementation here includes some more stability:
   - Concatenating two children types consisting only of strings has a more consistent result...
      - In master, we produce `[ 'First paragraph', 'Second paragraph' ]` where here we generate the simpler and consistent `[ 'First paragraphSecond paragraph' ]`
   - Empty node deletion traverses deeply, resolving more issues with zero-width space character presence (see included E2E test case)
- I wasn't entirely sure about RichText manually crafting its own BlockChildren objects, vs. something like with `dom-react` in passing a map/reduce callback (or object of "visitors") into the block API itself.
   - Many complications:
      - We may need to clean either TEXT_NODE and ELEMENT_NODE
      - We may need to omit nodes altogether
      - Doing so in a way that is performant (i.e. single pass)
   - Since this value is aimed at RichText, it seemed reasonable that it is fully aware of the object shape

**Testing instructions:**

Verify there are no regressions in the display, manipulation (splitting, merging, transforming), and save+reload of blocks leveraging RichText (e.g. paragraph, quote). A particular focus should be made on areas of low test coverage (e.g. block transforms).
